File: javascript/test/test_gilded_rose.js
const fs = require('fs');
const {assert} = require('chai');
const {Shop, Item} = require('../src/gilded_rose.js');

describe("Gilded Rose", function() {

  it("update quality of a normal item", function() {
    const foo = new Item("foo", 20, 10);

    const shop = new Shop([foo]);
    const items = shop.updateQuality();

    assert.equal(foo.sellIn, 19);
    assert.equal(foo.quality, 9);
  });

  specify('golden master', function() {
    const { Shop, Item } = require("../src/gilded_rose");

    const items = [
      new Item("+5 Dexterity Vest", 10, 20),
      new Item("Aged Brie", 2, 0),
      new Item("Elixir of the Mongoose", 5, 7),
      new Item("Sulfuras, Hand of Ragnaros", 0, 80),
      new Item("Sulfuras, Hand of Ragnaros", -1, 80),
      new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
      new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
      new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),

      // This Conjured item does not work properly yet
      new Item("Conjured Mana Cake", 3, 6),
    ];

    const actualLog = [];
    const days = 30;
    const gildedRose = new Shop(items);

    for (let day = 0; day <= days; day++) {
      actualLog.push(`-------- day ${day} --------`);
      actualLog.push("name, sellIn, quality");
      items.forEach(item => actualLog.push(`${item.name}, ${item.sellIn}, ${item.quality}`));
      gildedRose.updateQuality();
    }

    const text = fs.readFileSync("../golden-master/expected-output.txt", "UTF-8");
    const lines = text.split(/\r?\n/);
    const expectedLog = lines.filter(x => x.length != 0);

    assert.deepEqual(actualLog, expectedLog);

  });

})

File: python/test_gilded_rose.py
from gilded_rose import GildedRose, Item


def test_default_item():
    default_item = Item("default", sell_in=10, quality=20)
    items = [default_item]
    shop = GildedRose(items)

    shop.update_quality()

    assert default_item.sell_in == 9
    assert default_item.quality == 19


def test_gloden_master():
    items = [
        Item(name="+5 Dexterity Vest", sell_in=10, quality=20),
        Item(name="Aged Brie", sell_in=2, quality=0),
        Item(name="Elixir of the Mongoose", sell_in=5, quality=7),
        Item(name="Sulfuras, Hand of Ragnaros", sell_in=0, quality=80),
        Item(name="Sulfuras, Hand of Ragnaros", sell_in=-1, quality=80),
        Item(name="Backstage passes to a TAFKAL80ETC concert", sell_in=15, quality=20),
        Item(name="Backstage passes to a TAFKAL80ETC concert", sell_in=10, quality=49),
        Item(name="Backstage passes to a TAFKAL80ETC concert", sell_in=5, quality=49),
        Item(name="Conjured Mana Cake", sell_in=3, quality=6),
    ]

    gilded_rose = GildedRose(items)
    actual_log = []
    days = 30
    for day in range(days + 1):
        actual_log.append(f"-------- day {day} --------")
        actual_log.append("name, sellIn, quality")
        for item in items:
            actual_log.append(str(item))
        gilded_rose.update_quality()

    with open("../golden-master/expected-output.txt") as f:
        expected_log = f.read().splitlines(keepends=False)

    expected_log = [x for x in expected_log if x]

    assert actual_log == expected_log


if __name__ == "__main__":
    # In case pytest is not easy to install
    test_default_item()
    test_gloden_master()
